# Zero to Hero

## 1.09 Dividir en Training y Testing
## Version libreria caret

El objetivo es hacer una division del dataset en training/testing que sea estratificada en la clase
<br>
https://es.wikipedia.org/wiki/Muestreo_estratificado

In [21]:
options(warn=-1)

In [22]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria rpart
library( "rpart.plot") #cargo la libreria rpart.plot
library("caret") #cargo la libreria caret

#Aca tuve bastantes problemas para instalar desde Jupyter la libreria caret. Pude hacerlo desde el prompt de anaconda con "conda install -c r r-cart"

used    (Mb) gc trigger (Mb) max used (Mb) 
Ncells  826893 44.2 1470791    78.6  1470791  78.6
Vcells 1491908 11.4 9184328    70.1 34968824 266.8

In [23]:
options(repr.plot.width=20, repr.plot.height=10) 
setwd("C:\\Users\\oliva\\OneDrive\\Desktop\\DS\\Austral\\08 - Labo1") #Aqui se debe poner la ruta de la PC local

In [24]:
dataset <- fread("./labo/datasets/paquete_premium_202011.csv")   #cargo el dataset

https://www.rdocumentation.org/packages/caret/versions/6.0-88/topics/createDataPartition
<br>
la funcion  **createDataPartition** devolverá un vector de posiciones, las que cumplen con la particion indicada
<br>
*p= 0.5* significa que queremos el 50% de los registros
<br>
*dataset$clase_ternaria*   es el vector con la clase ternaria, que es por donde se estratificará
<br>

In [25]:
train_rows <- createDataPartition(dataset$clase_ternaria, p = 0.50, list = FALSE)

veamos que ha quedado en  los primeros registros del vector **train_rows**

In [26]:
head(train_rows)

Resample1
[1,]  1       
[2,]  3       
[3,]  6       
[4,]  7       
[5,]  8       
[6,] 10

Aqui ya vemos que la particion tiene un suconjunto de los registros

Ahora creamos los datasets de training y testing

In [28]:
dtrain <- dataset[train_rows]

-train_rows  significa el complemento  (no confundir con numeros negativos)

In [29]:
dtest <- dataset[-train_rows] #Es decir, me quedo con lo que no esta contenido en train_rows

Compruebo la division

In [30]:
nrow(dtrain)
nrow(dtest)
nrow(dtrain) + nrow(dtest)
nrow(dataset)

[1] 79695

[1] 79693

[1] 159388

[1] 159388

Compruebo que la particion es estratificada

In [31]:
dataset[  , .N, clase_ternaria] #todas las filas, por columna, agrupar por clase_ternaria
dtrain[  , .N, clase_ternaria]
dtest[  , .N, clase_ternaria]

clase_ternaria N     
1 CONTINUA       158229
2 BAJA+1            511
3 BAJA+2            648

clase_ternaria N    
1 CONTINUA       79115
2 BAJA+2           324
3 BAJA+1           256

clase_ternaria N    
1 CONTINUA       79114
2 BAJA+1           255
3 BAJA+2           324

La division se ha construido en forma estratificada

La cantidad de BAJA+2 es 648,  y se observa que training tiene 324 BAJA+2 ,exactamente la mitad de 648,  y testing tambien 324 BAJA+2